File: KMP.py
# https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/description/
# 找到s2在s1中第一个出现的下标

def build_next(s):
    m = len(s)
    if m == 1:
        return [-1]
    nxt = [-1] * m
    nxt[1] = 0
    i, cn = 2, 0
    while i < m:
        if s[cn] == s[i - 1]:
            nxt[i] = cn + 1
            i += 1
            cn += 1
        elif cn > 0:
            cn = nxt[cn]
        else:
            nxt[i] = 0
            i += 1
    return nxt

def kmp(s1, s2):
    n, m = len(s1), len(s2)
    nxt = build_next(s2)
    x = y = 0
    while x < n and y < m:
        if s1[x] == s2[y]:
            x += 1
            y += 1
        elif y == 0:
            x += 1
        else:
            y = nxt[y]
    return x - m if y == m else -1

File: KMP.cpp
#include <bits/stdc++.h>
using namespace std;

// https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/description/
// 找到s2在s1中第一个出现的下标

vector<int> build_next(string& s) {
  int m = s.size();
  if (m == 1) {
    return {-1};
  }
  vector<int> nxt(m, -1);
  nxt[1] = 0;
  int i = 2, cn = 0;
  while (i < m) {
    if (s[i - 1] == s[cn]) {
      nxt[i++] = ++cn;
    } else if (cn == 0) {
      nxt[i++] = 0;
    } else {
      cn = nxt[cn];
    }
  }
  return nxt;
}

int kmp(string& s1, string& s2) {
  int n = s1.size(), m = s2.size(), x = 0, y = 0;
  vector<int> nxt = build_next(s2);
  while (x < n && y < m) {
    if (s1[x] == s2[y]) {
      ++x;
      ++y;
    } else if (y == 0) {
      ++x;
    } else {
      y = nxt[y];
    }
  }
  return y == m ? x - y : -1;
}
